Updated Uri to prevent instigation with an invalid scheme

trowski opened the pull request:
Uri::applyParts() now validates the scheme so a Uri can not be instigated with an invalid scheme.

Changed Uri::getPort() to return null if the port is the default port for the scheme. See PSR-7 specification.

mtdowling commented:
I like it being able to accept any scheme as that is allowed by the current proposal (it uses a `MAY` to state that implementations can support additional schemes). Maybe we can update it to accept any scheme and strtolower() the scheme instead of forcing http and https. Also, filtering the scheme should not require a regular expression as the characters ":" and "/" are not allowed in the scheme so they can easily be removed using `rtrim($scheme, ':/')`.

I also prefer that calls to the accessor methods return a value that is already calculated if possible (i.e., calling getPort should just return a pre-calculated port value). Because these values are immutable, it's very easy to calculate a return value and store it internally for subsequent calls. This allows us to determine if it is a default port when the port is set, not when the port is retrieved.

trowski commented:
Updated the scheme filter to accept any scheme. I would pick one or the other, since it doesn't make sense to allow any scheme on instigation and then force http or https with `Uri::withScheme()`. Also removed the `preg_replace()` call to `rtrim($scheme, ':/')` as you suggested.

The port is now pre-calculated so calls to `Uri::getPort()` are lightweight.

mtdowling commented on src/Uri.php:
```diff
@@ -490,15 +492,47 @@ private static function isNonStandardPort($scheme, $host, $port)
+        $scheme = strtolower($scheme);
+        $scheme = rtrim($scheme, ':/');
+
+        if (empty($scheme)) {
```
Is there a need for this empty check?

mtdowling commented:
> I would pick one or the other, since it doesn't make sense to allow any scheme on instigation and then force http or https with Uri::withScheme().

Agreed.

trowski commented:
Oops, no. It was necessary when the scheme was validated, but not now.

mtdowling commented on src/Uri.php:
```diff
@@ -321,18 +326,15 @@ public function getFragment()
-                $scheme
-            ));
+        if ($this->scheme === $scheme) {
+            return clone $this;
```
this should just return `$this` not a clone because it didn't need any changes.

trowski commented:
I wasn't sure. phly/http returned clones even when nothing changed. Does the spec say that $this can be returned if nothing changes?

mtdowling commented:
Yeah, the spec was updated recently to show that you can return the same instance when no change is needed. I prefer to do that when possible to avoid the clone.

trowski commented:
I should quickly updated the other methods to do the same then.

trowski commented:
Is Uri::getPath() always suppose to return `'/'` if the path is empty?

mtdowling commented:
No that was recently removed as a restriction so that relative paths are now possible.

trowski commented:
Right, I actually deleted the question asking about it since I realized soon after posting that I was looking at an old version of the DocComments on `UriInterface`.

How does that last commit look?

mtdowling commented:
Looks good! Thanks, Aaron!
